test/Microsoft.Azure.Devices.Edge.Test.Common/EdgeModule.cs

```diff
@@ -17,25 +17,20 @@ public enum EdgeModuleStatus
     public class EdgeModule
     {
         protected string deviceId;
-        protected IotHub iotHub;
 
-        public EdgeModule(string id, string deviceId, IotHub iotHub)
+        public EdgeModule(string id, string deviceId)
         {
             this.deviceId = deviceId;
-            this.iotHub = iotHub;
             this.Id = id;
         }
 
         public string Id { get; }
 
         public static Task WaitForStatusAsync(EdgeModule[] modules, EdgeModuleStatus desired, CancellationToken token)
```

Comment:
Since module.deviceId is not considered in the method, should it be clearer with module ids as input.

Reply:
```suggestion
        public static Task WaitForStatusAsync(Hashset<string> moduleIds, EdgeModuleStatus desired, CancellationToken token)
```

Reply:
Should it be moved to EdgeDevice?

Reply:
With one of my recent commits do address another comment, it _does_ access `deviceId`.

Look at the public interface for the EdgeDaemon class: `InstallAsync`, `ConfigureAsync`, `StartAsync`, `StopAsync`, `UninstallAsync`, `WaitForStatusAsync`. That last method is about waiting for the status of the daemon itself. So all these methods have to do with the management of the daemon, not the running of modules (which is managed by the runtime).

Reply:
Can I understand like this?
EdgeDaemon is equivalent to iot edge service; and it provides functionalities like "iotedge list" or "iotedge version".  Therefore EdgeDaemon has knowledge about what modules are running in this Edge device.  Since the method has EdgeModule array argument, i think it is ok to be in EdgeDaemon.  Since it is a static method,  I am ok if you prefer this helper method in EdgeModule class.

Reply:
We're talking about slightly different things. The daemon does have knowledge about what modules are running, but the EdgeDaemon class in the end-to-end test library is focused on installation and status of the daemon, not installation and status of the things the daemon manages (modules).

What do you think about this:
A typical pattern in the tests is:
```csharp
  var seekTime = await runtime.DeployConfigurationAsync(b => b.AddModule("m1", image), token);
  var m1 = new EdgeModule("m1", deviceId);
  await runtime.WaitForModulesRunningAsync(new[] { m1 }, token);
  await m1.WaitForEventsReceivedAsync(seekTime, iotHub, token);
```

What if `EdgeRuntime.DeployConfigurationAsync` could automatically create and return an `EdgeModule` for each module in the deployment config? The new pattern would be simpler:

```csharp
  var (seekTime, modules) = await runtime.DeployConfigurationAsync(b => b.AddModules("m1", image), token);
  await runtime.WaitForModulesRunningAsync(modules.Values, token);
  await modules["m1"].WaitForEventsReceivedAsync(seekTime, iotHub, token);
```

That way, modules are always created with the right deviceId. Also, a little less code in the tests.

Reply:
Alright, I just added a commit that introduces the `EdgeDeployment` class, which represents the set of modules that were deployed to the edge device. It is returned from `EdgeRuntime.DeployConfigurationAsync` after the deployment has been sent to the cloud, the edge device has received it, and all the modules in the deployment are up and running. This simplifies the tests because they no longer have to create a config AND create an `EdgeModule` for each module in their deployment; they simply create a config, then the deploy method returns the set of `EdgeModules` they can use to listen for events, do twin operations, or send direct methods. (Note I also deleted `ModuleTwin.cs` and moved its functions into `EdgeModule`--the separation between twin operations and all other operations on a module is unnecessary). Another benefit is that tests often don't need to refer to the deviceId at all--the device was created during setup, so the tests themselves just focus on modules.

So the typical test pattern now looks like this:

```csharp
var deployment = await runtime.DeployConfigurationAsync(b => b.AddModule("m1", image), token);
var m1 = deployment.Modules["m1"];
await m1.WaitForEventsReceivedAsync(deployment.StartTime, token);
```